Feature/login context

airenzp opened the pull request:
Solves #277 #276, #273 and #313
This PR creates a context to access the user data in order to validate if the report can be accessed. Since the frontend user was kept in a state variable it was lost when refreshing the report, now the user state it is persisted in the browser to avoid this error.  
The default url is set now to /report and there is some refactoring to do the redirect to /login possible in the hooks.

jonandernovella commented:
Will test this one tomorrow

HenrikeW commented:
Will also look at this now.

HenrikeW commented on frontend/src/components/AuthProvider.tsx:
```diff
@@ -0,0 +1,84 @@
+  const [user, setUser] = React.useState();
+  const navigate = useNavigate();
+
+  React.useEffect(() => {
+    let user = window.localStorage.getItem("user");
+    if (user) user = JSON.parse(user);
+    setUser(user);
+  }, []);
+
+  React.useEffect(() => {
+    if (user) window.localStorage.setItem("user", JSON.stringify(user));
+  }, [user]);
```
You might need cleanup functions here? 
https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1

airenzp replied:
Thanks for the input Henrike. I added a commit to fix the error. The logout frontend logic should be out of the async call I think

airenzp replied:
This error happens because of the redirect to login. We need to cancel the refresh of the page when redirecting, which I am doing in my last commit. Still sometimes I see the error happening.

HenrikeW reviewed (changes requested):
I haven't looked at the code in detail yet, but saw two bugs (?) already when testing it:
1. When I'm logged out and go to localhost:4567 I see an empty Report page for about a second before I'm redirected to /login. 
2. I see an error in the console that I believe wasn't there before. 
```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
Report@http://localhost:4567/dist/pages/Report.js:17:51
```
I made a comment in the code where I guess something could cause the error, but I'm definitely not sure if that really is where it comes from.

HenrikeW commented:
The behavior is still the same as described in my previous comment, both concerning 1. and 2. 
The error I mentioned in 2. occurs when loading the login page. 
I also checked if the short appearance of the report page (described in 1) is a caching issue, but it's not. I see the same after cleared cache and in a private window.

airenzp commented:
> The behavior is still the same as described in my previous comment, both concerning 1. and 2. The error I mentioned in 2. occurs when loading the login page. I also checked if the short appearance of the report page (described in 1) is a caching issue, but it's not. I see the same after cleared cache and in a private window.

There is only one error I think, and it happens because when we redirect to login we need to cancel the refresh of the page when redirecting, which I am doing in my last commit. Still sometimes I see the error happening.

jonandernovella reviewed (changes requested):
i have done a test with the session time limited to one minute, and I am not seeing the behaviour i'd expect.

1) When I don't save changes, the session lasts for one minute, as expected
2) If I save changes during my session, the browser would still kick me out after one minute.

airenzp commented:
I added some commits with improvements and fixes in the code and rebased on develop to get the latest changes. The unmount error is happening now if you reload the page and do login/logout, not on every logout. I am suspecting that it is related with the previous error about the duplicated keys, since they happen together, only after reloading the page.

airenzp commented:
I created an issue about the duplicated keys and mount errors, see here #313. We can talk further next week if you want, just ping me.

airenzp commented:
> i have done a test with the session time limited to one minute, and I am not seeing the behaviour i'd expect.
> 
> 1. When I don't save changes, the session lasts for one minute, as expected
> 2. If I save changes during my session, the browser would still kick me out after one minute.

I noticed this when trying to hack the bakend to force the session to expire. You can also try to logout from another tab and use the current tab to test this PR.

jonandernovella commented on frontend/src/components/AuthProvider.tsx:
```diff
@@ -0,0 +1,81 @@
+      })
+      .catch((error) => console.log("An error occured.", error));
+
+    setUser(user);
+    return user;
```
It would be nice if we keep this outside of this function.

jonandernovella commented:
I spotted two things that I think we should fix after merging this PR See #349 and #348

jonandernovella reviewed (approved):
Looking good! There are some minor things to be addressed which are noew issues.
